src/framework/audio/internal/worker/audioiohandler.cpp

```diff
@@ -0,0 +1,165 @@
+
+#include "log.h"
+#include "async/async.h"
+
+#include "internal/audiosanitizer.h"
+#include "internal/audiothread.h"
+#include "audioerrors.h"
+
+using namespace mu::audio;
+using namespace mu::async;
+
+AudioIOHandler::AudioIOHandler(IGetTrackSequence* getSequence)
+    : m_getSequence(getSequence)
+{
+}
+
+Promise<AudioInputParams> AudioIOHandler::inputParams(const TrackSequenceId sequenceId, const TrackId trackId) const
+{
+    return Promise<AudioInputParams>([this, sequenceId, trackId](Promise<AudioInputParams>::Resolve resolve,
+                                                                 Promise<AudioInputParams>::Reject reject) {
+        ONLY_AUDIO_WORKER_THREAD;
+
+        ITrackSequencePtr s = sequence(sequenceId);
+
+        if (!s) {
+            reject(static_cast<int>(Err::InvalidSequenceId), "invalid sequence id");
+        }
+
+        RetVal<AudioInputParams> result = s->audioIO()->inputParams(trackId);
+
+        if (!result.ret) {
+            reject(result.ret.code(), result.ret.text());
+        }
+
+        resolve(result.val);
+    }, AudioThread::ID);
+}
+
+void AudioIOHandler::setInputParams(const TrackSequenceId sequenceId, const TrackId trackId, const AudioInputParams& params)
+{
+    Async::call(this, [this, sequenceId, trackId, params]() {
+        ONLY_AUDIO_WORKER_THREAD;
+
+        ITrackSequencePtr s = sequence(sequenceId);
+
+        if (s) {
+            s->audioIO()->setInputParams(trackId, params);
+        }
+    }, AudioThread::ID);
+}
+
+Channel<TrackSequenceId, TrackId, AudioInputParams> AudioIOHandler::inputParamsChanged() const
+{
+    return m_inputParamsChanged;
+}
+
+Promise<AudioOutputParams> AudioIOHandler::outputParams(const TrackSequenceId sequenceId, const TrackId trackId) const
+{
+    return Promise<AudioOutputParams>([this, sequenceId, trackId](Promise<AudioOutputParams>::Resolve resolve,
+                                                                  Promise<AudioOutputParams>::Reject reject) {
+        ONLY_AUDIO_WORKER_THREAD;
+
+        ITrackSequencePtr s = sequence(sequenceId);
+
+        if (!s) {
+            reject(static_cast<int>(Err::InvalidSequenceId), "invalid sequence id");
+        }
+
+        RetVal<AudioOutputParams> result = s->audioIO()->outputParams(trackId);
+
+        if (!result.ret) {
+            reject(result.ret.code(), result.ret.text());
+        }
+
+        resolve(result.val);
+    }, AudioThread::ID);
+}
+
+void AudioIOHandler::setOutputParams(const TrackSequenceId sequenceId, const TrackId trackId, const AudioOutputParams& params)
+{
+    Async::call(this, [this, sequenceId, trackId, params]() {
+        ONLY_AUDIO_WORKER_THREAD;
+
+        ITrackSequencePtr s = sequence(sequenceId);
+
+        if (s) {
+            s->audioIO()->setOutputParams(trackId, params);
+        }
+    }, AudioThread::ID);
+}
+
+Channel<TrackSequenceId, TrackId, AudioOutputParams> AudioIOHandler::outputParamsChanged() const
+{
+    return m_outputParamsChanged;
```

Comment:
I believe we need to add a thread validation/indication marker:
* ONLY_AUDIO_WORKER_THREAD
* ONLY_AUDIO_MAIN_THREAD
* ONLY_AUDIO_MAIN_OR_WORKER_THREAD
* ANY_AUDIO__THREAD

Reply:
yeah, makes sense

added in https://github.com/musescore/MuseScore/pull/8471/commits/1f4270c0615b7641b18b9f43c8f0026b35f7392c